src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -6117,16 +6118,23 @@ private void LoadSearchValues(ReadOnlySpan<char> chars)
             // Logically do _searchValues[index], but avoid the bounds check on accessing the array,
             // and cast to the known derived sealed type to enable devirtualization.
 
-            // DerivedSearchValues d = Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(this._searchValues), index);
+            // DerivedSearchValues d = Unsafe.As<DerivedSearchValues>(Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(this._searchValues), index));
             // ... = d;
             Ldthisfld(s_searchValuesArrayField);
             Call(s_memoryMarshalGetArrayDataReferenceSearchValues);
             Ldc(index * IntPtr.Size);
             Add();
             _ilg!.Emit(OpCodes.Ldind_Ref);
-            LocalBuilder ioavLocal = _ilg!.DeclareLocal(list[index].GetType());
-            Stloc(ioavLocal);
-            Ldloc(ioavLocal);
+            Call(MakeUnsafeAs(list[index].GetType())); // provide JIT with details necessary to devirtualize calls on this instance
+
+            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2060:MakeGenericMethod", Justification =
+                "Only uses Unsafe.As<T> with reference types <T>, which it also uses in this method to ensure it's rooted. Unsafe.As is an intrinsic with a nop " +
+                "implementation supplied at codegen time. And this is also part of the regex compiler that's only used when dynamic code compilation is supported.")]
+            static MethodInfo MakeUnsafeAs(Type type)
+            {
+                Debug.Assert(!type.IsValueType);
```

Comment:
```suggestion
            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2060:MakeGenericMethod", Justification =
                "Calling Unsafe.As<T> is safe since the T doesn't have trimming annotations.")]
            static MethodInfo MakeUnsafeAs(Type type)
            {
```

Reply:
In the other reply (https://github.com/dotnet/runtime/pull/86010#discussion_r1190038522) I explained what the current warning means. (I added this comment so I could suggest the justification message and remove the Assert). The current warning is a "trimming IL" warning, not an AOT warning.

The warning you get for Native AOT when using MakeGenericType/Method is `IL3050` - https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/warnings/il3050

Reply:
> The current warning is a "trimming IL" warning

Ok. I see the details now in https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trim-warnings/il2060.

> [RequiresDynamicCode]

It's strange to me that we lump MakeGenericMethod in as part of RequiresDynamicCodeAttribute.  There's a meaningful difference between something that uses reflection emit and must be able to compile (or interpret) at execution time, and something that just needs the code to be available.  MakeGenericMethod doesn't require dynamic code if you only use instantiations that are known to already be available.  The JIT can create such code on-demand, but in an AOT model where all the code is created ahead of time, it's not that MakeGenericMethod requires dynamic code, it's that it requires the code to have already been compiled, and it fails if it wasn't.

Reply:
> It's strange to me that we lump MakeGenericMethod in as part of RequiresDynamicCodeAttribute.

The way I think of it is that MakeGenericMethod has correctness concerns for both trimming and when not being able to JIT code (i.e. Native AOT). So when you use it, you will typically get 2 warnings - one for trimming (IL2060) and one for AOT (IL3050). The IL3050 warning is automatically suppressed here because this code already says it requires dynamic code, so there's no point in warning about it.

But the trimming warning IL2060 is still emitted because this code can be used in a trimmed app. Even if we never did Native AOT, you'd still get the warning - but like I said in the other [thread](https://github.com/dotnet/runtime/pull/86010#discussion_r1190038522), I think it may be a bug in the trimmer here because it should be able to statically see that there are no trimming annotations on Unsafe.As.

Reply:
> MakeGenericMethod doesn't require dynamic code if you only use instantiations that are known to already be available. The JIT can create such code on-demand, but in an AOT model where all the code is created ahead of time, it's not that MakeGenericMethod requires dynamic code, it's that it requires the code to have already been compiled, and it fails if it wasn't.

The part that it is warning about is

> and it fails if it wasn't.

It can't statically guarantee what you are doing is going to work. So it warns.

Note that "RequiresDynamicCode" isn't restricted to just generating IL code. It can also mean that it needs the runtime to generate native code as well. That's why things like `Array.CreateInstance(Type elementType, int length)` are annotated with `RequiresDynamicCode`.

Reply:
> Note that "RequiresDynamicCode" isn't restricted to just generating IL code. It can also mean that it needs the runtime to generate native code as well. That's why things like Array.CreateInstance(Type elementType, int length) are annotated with RequiresDynamicCode.

I understand.  I'm saying that there's a difference between "this will need the runtime to generate native code" and "this may need the runtime to generate native code".

Reply:
In either case, it still needs to warn, right? If it can't tell that you've structured things well, it has to inform you that it is possible your code may not work.

Reply:
Yes. My comment isn't about warning vs not. It's about naming and communication of what's required and what to do about it.

Reply:
Agreed, we don't have a great story about when it is safe to suppress these warnings and when it isn't. Maybe the above docs should be updated with "When to suppress warnings" sections when applicable. Like we do for other analyzer warnings. Ex.

 https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1033#when-to-suppress-warnings

Reply:
I'll let @MichalStrehovsky correct me, but I think `MakeGenericMethod` is AOT incompatible because it might need runtime code gen. If all of the generic arguments are reference types, then it should not need code gen, but if one of them is a value type, it needs to have the code for it pre-generated. Kind of similar to `List<string>` and `List<object>` will share the same code and thus you can at runtime create a new `List<MyRefType>`. But `List<int>` and `List<bool>` don't share code, so you can't create a `List<double>` at runtime (unless it was compiled into the app).
To my understanding that is the reason why `MakeGenericMethod` can produce `IL3050`.

That is all assuming it can statically figure out which generic method is being instantiated, if it can't even tell that, then it should warn always.

Reply:
Yes, understood.  Maybe I'm splitting hairs here, but to me "requires dynamic code" suggests it will _always_ require dynamic code, and that's not the case.  If you have `SomeMethod<T>` and you use `MakeGenericMethod(typeof(double))`, that should work just fine as long as your app includes the code for `SomeMethod<double>`.  This is really `[NeedsPrecompiledCode]`, such that it will fail if such code isn't available.

Reply:
That's true (I will restrain from discussing the attribute name ;-)). In theory the compiler could figure lot of this out and warn much less (it can know that it generated code for `SomeMethod<double>`, so it doesn't need to warn if it sees it being constructed via `MakeGenericMethod`). I'm definitely open to suggestions on which parts of this are more important to improve on (make the compiler more clever).

There's a question of how much we want to make the compiler clever in general though since the analyzer will not be able to do the same - and then deciding what is the right balance between diagnostics from the analyzer and the compiler.

Reply:
The `RequiresDynamicCode` work similar to `RequiresUnreferencedCode`. Both of them can work under some circumstances and that's why it's sometimes possible to suppress the warning.

The fact that RDC on `MakeGenericType` is sometimes suppressible and RDC on Ref.Emit is never suppressible is a thing that could change. E.g. if there's enough of a business justification, we could recognize that some Ref.Emit uses are just about building "fast field accessors" and intrinsically convert them to the fast field accessors we're planning to add. Then such Ref.Emit use is also technically safe and suppressible.

The attribute really only captures that "unless otherwise specified, this will require runtime code generation".

Reply:
> The attribute really only captures that "unless otherwise specified, this will require runtime code generation".

That's what I'm struggling with, from a naming/messaging perspective. There is no runtime code generation with native aot, so saying something requires runtime code generation to work suggests it'll never work on native aot, but obviously that's not the case. The real condition in this case isn't that it "requires runtime code generation" but rather that it "requires certain code to be available", which in the case of native aot means it had to generate that code at build time. 

Anyway, I'll drop it. I think we're making this confusing, though.